src/lib/gossip_net/gossip_net.ml

```diff
@@ -239,26 +264,59 @@ module Make (Message : Message_intf) :
-                    (* connection state is the client's IP and ephemeral port
+              let conn_map =
+                Option.value_map
+                  ~default:(Hashtbl.create (module Uuid))
+                  (Hashtbl.find t.connections (Socket.Address.Inet.addr client))
+                  ~f:Fn.id
+              in
+              if
+                Option.is_some t.max_concurrent_connections
+                && Hashtbl.length conn_map
+                   = Option.value_exn t.max_concurrent_connections
+              then (
+                Logger.error t.logger ~module_:__MODULE__ ~location:__LOC__
+                  "Cannot open another connection. Number of open connections \
+                   from %s equals the limit %d."
+                  (Socket.Address.Inet.to_string client)
+                  (Option.value_exn t.max_concurrent_connections) ;
+                Reader.close reader >>= fun _ -> Writer.close writer )
+              else
+                let conn_id = Uuid.create () in
+                let%map _ =
+                  Rpc.Connection.server_with_close reader writer
+                    ~implementations
+                    ~connection_state:(fun conn ->
+                      (* connection state is the client's IP and ephemeral port
                         when connecting to the server over TCP; the ephemeral
                         port is distinct from the client's discovery and
                         communication ports *)
-                    Hashtbl.set t.connections
-                      ~key:(Socket.Address.Inet.addr client)
-                      ~data:conn ;
-                    Socket.Address.Inet.to_host_and_port client )
-                  ~on_handshake_error:
-                    (`Call
-                      (fun exn ->
-                        Logger.error t.logger ~module_:__MODULE__
-                          ~location:__LOC__ "%s" (Exn.to_string_mach exn) ;
-                        Deferred.unit ))
-              in
-              Hashtbl.remove t.connections @@ Socket.Address.Inet.addr client
-              )
+                      let () =
+                        Hashtbl.add_exn conn_map ~key:conn_id ~data:conn
+                      in
+                      Hashtbl.set t.connections
+                        ~key:(Socket.Address.Inet.addr client)
+                        ~data:conn_map ;
+                      Socket.Address.Inet.to_host_and_port client )
+                    ~on_handshake_error:
+                      (`Call
+                        (fun exn ->
+                          Logger.error t.logger ~module_:__MODULE__
+                            ~location:__LOC__ "%s" (Exn.to_string_mach exn) ;
+                          Deferred.unit ))
+                in
+                let conn_map =
+                  Hashtbl.find_exn t.connections
+                    (Socket.Address.Inet.addr client)
+                in
+                let () = Hashtbl.remove conn_map conn_id in
```

Comment:
don't need a let binding here either

Reply:
I find let bindings to be more readable especially when there is one expression ending with a `;` in a series of let bindings